ipa-core/src/protocol/ipa_prf/malicious_security/verifier.rs

```diff
@@ -370,4 +401,30 @@ mod test {
 
         assert_eq!(q_final_another_way.as_u128(), EXPECTED_Q_FINAL);
     }
+
+    #[test]
+    fn differences_are_zero() {
+        const ZKP_1_and_2_LEFT: [u128; 14] = [0, 0, 13, 17, 11, 25, 7, 11, 25, 17, 9, 22, 23, 3];
+        const ZKP_1_and_2_RIGHT: [u128; 14] = [0, 30, 16, 13, 25, 3, 6, 1, 12, 29, 30, 7, 7, 3];
+        const OUT_LEFT: u128 = 27;
+        const OUT_RIGHT: u128 = 0;
+
+        let zkp_left = make_chunks::<Fp31, 7>(&ZKP_1_and_2_LEFT);
+        let zkp_right = make_chunks::<Fp31, 7>(&ZKP_1_and_2_RIGHT);
+        let challenges = vec![Fp31::truncate_from(22u128), Fp31::truncate_from(17u128)];
+
+        let g_differences_left = compute_g_differences::<_, 7, 4>(&zkp_left, &challenges);
+        let g_differences_right = compute_g_differences::<_, 7, 4>(&zkp_right, &challenges);
+
+        let mut g_differences = g_differences_left
+            .iter()
+            .zip(g_differences_right)
+            .map(|(left, right)| *left + right)
+            .collect::<Vec<_>>();
+
+        g_differences[0] -= Fp31::truncate_from(OUT_LEFT) + Fp31::truncate_from(OUT_RIGHT);
```

Comment:
I would rather not do this here, but instead do it within the function itself.  Basically things the check for the first stage that it equals -m/2.